icu4c/source/common/locavailable.cpp

```diff
@@ -95,85 +97,170 @@ U_NAMESPACE_USE
 
 /* ### Constants **************************************************/
 
-/* These strings describe the resources we attempt to load from
- the locale ResourceBundle data file.*/
-static const char _kIndexLocaleName[] = "res_index";
-static const char _kIndexTag[]        = "InstalledLocales";
+namespace {
 
-static char** _installedLocales = NULL;
-static int32_t _installedLocalesCount = 0;
-static icu::UInitOnce _installedLocalesInitOnce;
+// Enough capacity for the two lists in the res_index.res file
+const char** gAvailableLocaleNames[2] = {};
+int32_t gAvailableLocaleCounts[2] = {};
+icu::UInitOnce ginstalledLocalesInitOnce = U_INITONCE_INITIALIZER;
 
-/* ### Get available **************************************************/
+class AvailableLocalesSink : public ResourceSink {
+  public:
+    void put(const char *key, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) U_OVERRIDE {
+        ResourceTable resIndexTable = value.getTable(status);
+        if (U_FAILURE(status)) {
+            return;
+        }
+        for (int32_t i = 0; resIndexTable.getKeyAndValue(i, key, value); ++i) {
+            int32_t type;
+            if (uprv_strcmp(key, "InstalledLocales") == 0) {
+                type = 0;
```

Comment:
In Java you are using the enum constants. Why not here?

Reply:
I was, until I changed the enum to having a third type, in which case the fields no longer correspond to each element in the enum.  I was considering changing Java but liked the convenience of the EnumMap.

I added an assertion to Java.

Reply:
Done as requested.